front/pages/api/w/[wId]/monthly-usage.ts

```diff
@@ -0,0 +1,159 @@
+      ) {
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message:
+              "The `referenceDate` query parameter is missing or invalid.",
+          },
+        });
+      }
+      const referenceDate = new Date(req.query.referenceDate);
+      const csvData = await getMonthlyUsage(referenceDate, owner.sId);
+      res.setHeader("Content-Type", "text/csv");
+      res.setHeader(
+        "Content-Disposition",
+        `attachment; filename=dust_monthly_usage_${referenceDate.getFullYear()}_${
+          referenceDate.getMonth() + 1
+        }.csv`
+      );
+      res.status(200).send(csvData);
+      return;
+
+    default:
+      return apiError(req, res, {
+        status_code: 405,
+        api_error: {
+          type: "method_not_supported_error",
+          message: "The method passed is not supported, GET is expected.",
+        },
+      });
+  }
+}
+
+export default withLogging(handler);
+
+async function getMonthlyUsage(
+  referenceDate: Date,
+  wId: string
+): Promise<string> {
+  const results = await front_sequelize.query<QueryResult>(
+    `
+    SELECT 
+      TO_CHAR(m."createdAt"::timestamp, 'YYYY-MM-DD HH24:MI:SS') AS "createdAt",
+      c."id" AS "conversationModelId",
+      m."id" as "MessageId",
+      um."id" AS "UserMessageId", 
+      am."id" AS "AgentMessageId",
+      u."id" as "UserId",
+      um."userContextFullName" AS "UserFullName",
+      COALESCE(ac."sId", am."agentConfigurationId") AS "AssistantId", 
+      COALESCE(ac."name", am."agentConfigurationId") AS "AssistantName", 
+      CASE 
+          WHEN ac."retrievalConfigurationId" IS NOT NULL THEN 'retrieval'
+          WHEN ac."dustAppRunConfigurationId" IS NOT NULL THEN 'dustAppRun'
+          ELSE NULL
+      END AS "ActionType",
+      CASE 
+          WHEN um."id" IS NOT NULL THEN 
+              CASE 
+                  WHEN um."userId" IS NOT NULL THEN 'web'
+                  ELSE 'slack'
+              END
+      END AS "source"
+  FROM 
+      "messages" m
+  JOIN 
+      "conversations" c ON m."conversationId" = c."id"
+  JOIN 
+      "workspaces" w ON c."workspaceId" = w."id"
+  LEFT JOIN 
+      "user_messages" um ON m."userMessageId" = um."id"
+  LEFT JOIN 
+      "users" u ON um."userId" = u."id"
+  LEFT JOIN 
+      "agent_messages" am ON m."agentMessageId" = am."id"
+  LEFT JOIN 
+      "agent_configurations" ac ON am."agentConfigurationId" = ac."sId" AND am."agentConfigurationVersion" = ac."version"
+  WHERE 
+      w."sId" = :wId AND
+      DATE_TRUNC('month', m."createdAt") = DATE_TRUNC('month', :referenceDate::timestamp)
+  ORDER BY 
+      "createdAt" DESC
+  `,
+    {
+      replacements: {
+        wId,
+        referenceDate: `${referenceDate.getFullYear()}-${
+          referenceDate.getMonth() + 1
+        }-${referenceDate.getDate()}`,
+      },
+      type: QueryTypes.SELECT,
+    }
+  );
+
+  const csvContent = results
```

Comment:
Need to handle case when there's no row?

Reply:
Can we add a header line?

Reply:
nvm me LGTM